File: 卡码/代码随想录/1/09-动态规划/Python/02-背包问题/01-01背包/LeetCode_0416.py
#416. 分割等和子集
from typing import *

class Solution:
  def canPartition(self, nums: List[int]) -> bool:
    sum:int=0
    for num in nums:
      sum+=num
    if sum%2==1:
      return False
    target:int=sum//2
    dp:List[int]=[0]*10001
    for i in range(len(nums)):
      for j in range(target,nums[i]-1,-1):
        dp[j]=max(dp[j],dp[j-nums[i]]+nums[i])
    if target==dp[target]:
      return True
    return False

File: 卡码/代码随想录/C++/动态规划/dp/02-背包问题/src/01-01背包/LeetCode_0416.cpp
//416. 分割等和子集
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	bool canPartition(vector<int>& nums) {
		int sum=0;
		for(int num:nums) sum+=num;
		if(sum%2==1) return false;
		int target=sum/2;
		vector<int> dp(10001,0);
		for(int i=0;i<nums.size();i++){
			for(int j=target;j>=nums[i];j--){
				dp[j]=max(dp[j],dp[j-nums[i]]+nums[i]);
			}
		}
		if(target==dp[target]) return true;
		return false;
	}
};

int main(){
  system("pause");
  return 0;
}
